packages/flutter/lib/src/widgets/fade_in_image.dart

```diff
@@ -7,17 +7,48 @@ import 'dart:typed_data';
 import 'package:flutter/foundation.dart';
 import 'package:flutter/services.dart';
 
+import 'animated_fade_out_fade_in.dart';
 import 'basic.dart';
 import 'framework.dart';
 import 'image.dart';
-import 'implicit_animations.dart';
-import 'transitions.dart';
+
 
 // Examples can assume:
 // Uint8List bytes;
 
+/// Signature for the [FadeInImage.animationBuilder] callback.
+///
+/// The `target` is the image that should be (eventually) shown when
+/// `isTargetLoaded == true`. `target` maps to [FadeInImage.image]. 
+/// 
+/// The `placeholder` is the image initially shown when `isTargetLoaded == false`.
+/// `placeholder` maps to [FadeInImage.placeholder].
+/// 
+/// `isTargetLoaded` will initially start as `false` and will transition to
+/// `true` when `target` supplies its first frame. It will transition back to
+/// `false` only if [FadeInImage.image] is changed to a new [ImageProvider].
+///
+///
+/// {@tool snippet}
+///
+/// ```dart
+/// Widget fadeOutFadeInAnimationBuilder(Widget target, Widget placeholder, bool isTargetLoaded) => 
+///   AnimatedFadeOutFadeIn(
+///     target: target,
+///     placeholder: placeholder,
+///     isTargetLoaded: isTargetLoaded,
+///     fadeInDuration: const Duration(milliseconds: 300),
+///     fadeOutDuration: Curves.easeOut,
+///     fadeInCurve: const Duration(milliseconds: 700),
+///     fadeOutCurve: Curves.easeIn,
+///   );
+/// ```
+/// {@end-tool}
+typedef FadeInImageAnimationBuilder = Widget Function(Widget target, Widget placeholder, bool isTargetLoaded);
```

Comment:
This doesn't really seem like a great abstraction level. You still have to implement (and presumably duplicate) a lot of the animation logic in your `FadeInImageAnimationBuilder`s.

Reply:
The thought is to delegate out the animation in a way that allows you to parameterise the animation behaviour without impacting the existing signature of `FadeInImage`.

E.g.

```dart
FadeInImage(
    placeholder: MemoryImage(bytes),
    image: NetworkImage('https://backend.example.com/image.png'),
    animationBuilder: (Widget target, Widget placeholder, bool isTargetLoaded) => AnimatedFadeOutFadeIn(
        target: target,
        placeholder: placeholder,
        isTargetLoaded: isTargetLoaded,
        fadeInDuration: const Duration(milliseconds: 300),
        fadeOutDuration: Curves.easeOut,
        fadeInCurve: const Duration(milliseconds: 700),
        fadeOutCurve: Curves.easeIn,
    );
)
```

or

```dart
FadeInImage(
    placeholder: MemoryImage(bytes),
    image: NetworkImage('https://backend.example.com/image.png'),
    animationBuilder: (Widget target, Widget placeholder, bool isTargetLoaded) => AnimatedCrossFade(
        firstChild: placeholder,
        secondChild: target,
        crossFadeState: isTargetLoaded ? CrossFadeState.showFirst : CrossFadeState.showSecond,
        duration: const Duration(milliseconds: 500),
    );
)
```

This would allow any consumer-written animations to be supported by the class without `FadeInImage` needing to know about the implementation details.